crates/bevy_app/src/app.rs

```diff
@@ -823,19 +829,37 @@ impl App {
     /// # }
     /// App::new().add_plugin(bevy_log::LogPlugin::default());
     /// ```
+    ///
+    /// # Panics
+    ///
+    /// Panics if the plugin was already added to the application.
     pub fn add_plugin<T>(&mut self, plugin: T) -> &mut Self
     where
         T: Plugin,
     {
-        self.add_boxed_plugin(Box::new(plugin))
+        match self.add_boxed_plugin(Box::new(plugin)) {
+            Ok(app) => app,
+            Err(AppError::DuplicatePlugin { plugin_name }) => panic!(
+                "Error adding plugin {}: : plugin was already added in application",
+                plugin_name
+            ),
+        }
     }
 
     /// Boxed variant of `add_plugin`, can be used from a [`PluginGroup`]
-    pub(crate) fn add_boxed_plugin(&mut self, plugin: Box<dyn Plugin>) -> &mut Self {
+    pub(crate) fn add_boxed_plugin(
+        &mut self,
+        plugin: Box<dyn Plugin>,
+    ) -> Result<&mut Self, AppError> {
         debug!("added plugin: {}", plugin.name());
+        if plugin.is_unique() && !self.plugin_name_added.insert(plugin.name().to_string()) {
```

Comment:
```suggestion
        if !self.plugin_name_added.insert(plugin.name().to_string()) && plugin.is_unique() {
```
We want to insert the name of the plugin even if it's not intended to be unique for debugging and editor purposes.

Reply:
there's already the `plugin_registry` which is a vec of all added plugins that is much better for that purpose

Reply:
Ah got it. I'm convinced.